File: 011-fibd/fibd.py
# Mortal Fibonacci Rabbits

# Sample Dataset
# 6 3
# Sample Output
# 4

def fibd_rabbits(n, m):
  if n < 0:
    return 0
  elif n < 3:
    return 1
  else:
    return fibd_rabbits(n-1, m) + fibd_rabbits(n-2, m) - fibd_rabbits(n-1-m, m)

File: 011-fibd/fibd.rb
# Mortal Fibonacci Rabbits

# Sample Dataset
# 6 3
# Sample Output
# 4

def fibd_rabbits(n, m)
  fibd = [1, 1]
  (3..n).each do |i|
    if i == m + 1
      fibd.append(fibd[-1] + fibd[-2] - 1)
    elsif i > m + 1
      fibd.append(fibd[-1] + fibd[-2] - fibd[-(m+1)])
    else
      fibd.append(fibd[-1] + fibd[-2])
    end
  end

  return fibd[-1]
end